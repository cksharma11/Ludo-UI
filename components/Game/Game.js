import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import CoinContainer from '../CoinContainer/CoinContainer';
import GameHeader from '../GameHeader/GameHeader';
import CellRow from '../CellRow/CellRow';
import GameStyles from './Game.style';
import Player from '../Player/Player';
import useGameHook from './GameHook';
import { startingCells } from '../../config/data/cells-structure';
import { API_URL } from '../../utils/utils';
import app from '../../https/app';
import constants from '../../constants/contants';
import CongratulationsPopup from '../CongratulationsPopup/CongratulationsPopup';
import colors from '../../config/styles/colors';

const { POLLING_INTERVAL } = constants;

const Game = ({ gameData: initialGameData, playerId, gameId }) => {
  const [gameData, setGameData] = useState(initialGameData);

  useEffect(() => {
    setTimeout(async () => {
      const res = await app.post(`${API_URL}/getGameData`, {
        body: JSON.stringify({ gameId })
      });
      setGameData(res);
    }, POLLING_INTERVAL);
  }, [gameData]);

  const {
    currentPlayerIndex,
    rollDice,
    diceValue,
    windowPlayer,
    getInactiveCoins,
    getCoins,
    isWindowPlayer,
    cellRowConfigs
  } = useGameHook({
    gameData,
    playerId
  });
  const {
    players: [player1, player2, player3, player4]
  } = gameData;

  const renderPlayer = (player, alignment, playerIndex) => {
    return player ? (
      <Player
        player={player}
        turn={currentPlayerIndex === playerIndex}
        alignment={alignment}
        onRollDice={rollDice}
        diceValue={diceValue}
        isWindowPlayer={isWindowPlayer(player, windowPlayer)}
      />
    ) : null;
  };

  const renderCoinContainer = (player, color) => {
    return (
      <CoinContainer
        nutralCoins={getInactiveCoins(getCoins(player))}
        color={color}
        showCoins={player !== undefined}
        gameData={gameData}
        playerId={playerId}
      />
    );
  };

  const renderCellRow = (index) => {
    const { className, color, containerClass } = cellRowConfigs[index];
    return (
      <CellRow
        className={className}
        cellId={startingCells[index]}
        color={color}
        containerClass={containerClass}
        gameData={gameData}
        playerId={playerId}
      >
        <style jsx>{GameStyles}</style>
      </CellRow>
    );
  };

  return (
    <div className="body">
      <GameHeader playerName={windowPlayer.name} gameId={gameData.id} />
      <section className="main_container">
        <section className="player_row">
          {renderPlayer(player1, 'left', 0)}
          {renderCoinContainer(player1, 'red')}
          {renderCellRow(0)}
          {renderCoinContainer(player2, 'blue')}
          {renderPlayer(player2, 'right', 1)}
        </section>

        <div className="board_middle_row">
          {renderCellRow(1)}
          <div className="clear_coins_container" />
          {renderCellRow(2)}
        </div>

        <section className="player_row">
          {renderPlayer(player3, 'left', 2)}
          {renderCoinContainer(player3, 'green')}
          {renderCellRow(3)}
          {renderCoinContainer(player4, 'yellow')}
          {renderPlayer(player4, 'right', 3)}
        </section>
      </section>
      {gameData.winningPlayer && (
        <CongratulationsPopup player={gameData.winningPlayer} />
      )}
      <style jsx global>{`
        body,
        html {
          @import url('https://fonts.googleapis.com/css?family=Open+Sans:300&display=swap');
          padding: 0;
          margin: 0;
          background: ${colors.palette1.shade5};
          font-family: 'Open Sans', sans-serif;
        }
        button {
          font-family: 'Open Sans', sans-serif;
          border-radius: 3px;
        }
      `}</style>
      <style jsx>{GameStyles}</style>
    </div>
  );
};

Game.propTypes = {
  gameData: PropTypes.object.isRequired,
  playerId: PropTypes.oneOfType([PropTypes.number, PropTypes.string])
    .isRequired,
  gameId: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired
};

export default Game;
