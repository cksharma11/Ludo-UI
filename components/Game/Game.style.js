import css from 'styled-jsx/css';
import colors from '../../config/styles/colors';

const GameStyles = css`
  .body {
    width: 100%;
    height: 100%;
    padding: 0;
    margin: 0;
    font-family: 'Trebuchet MS', 'Lucida Sans Unicode', 'Lucida Grande',
      'Lucida Sans', Arial, sans-serif;
  }

  .main_container {
    display: flex;
    padding: 15px;
    align-items: center;
    background: ${colors.palette1.shade2};
    flex-direction: column;
  }

  .player_row {
    display: flex;
  }

  .board_middle_row {
    display: flex;
    width: 630px;
    justify-content: space-between;
  }

  .coin_symbol {
    height: 50%;
    width: 50%;
    align-self: center;
    border: 1px solid #f1b091eb;
    border-radius: 50%;
  }

  .clear_coins_container {
    width: 0;
    height: 0;
    border-top: 63px solid rgb(84, 84, 235);
    border-bottom: 63px solid rgb(112, 236, 112);
    border-left: 63px solid rgb(231, 96, 96);
    border-right: 63px solid rgb(230, 230, 91);
  }
`;

export default GameStyles;
