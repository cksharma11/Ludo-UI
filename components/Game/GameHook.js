import { useState } from 'react';

const GameHook = ({ gameData: initialGameData }) => {
  const numberOfPlayer = initialGameData.players.length;
  const [currentPlayerIndex, setCurrentPlayerIndex] = useState(0);
  const [diceValue, setDiceValue] = useState(1);

  const updateCurrentPlayerIndex = () => {
    setCurrentPlayerIndex((currentPlayerIndex + 1) % numberOfPlayer);
  };

  const getDiceValue = () => {
    return Math.ceil(Math.random() * 6);
  };

  const rollDice = () => {
    setDiceValue(getDiceValue());
  };

  return {
    updateCurrentPlayerIndex,
    currentPlayerIndex,
    rollDice,
    diceValue
  };
};

export default GameHook;
