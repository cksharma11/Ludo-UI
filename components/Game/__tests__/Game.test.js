import React from 'react';
import { mount } from 'enzyme';
import Game from '../Game';

describe('Game', () => {
  const gameData = {
    players: [
      {
        playerName: 'test1',
        turn: true,
        playerId: 1
      },
      {
        playerName: 'test2',
        turn: false,
        playerId: 2
      },
      {
        playerName: 'test3',
        turn: false,
        playerId: 3
      },
      {
        playerName: 'test4',
        turn: false,
        playerId: 4
      }
    ]
  };

  it('should render properly', () => {
    const wrapper = mount(<Game gameData={gameData} playerId={1} />);
    expect(wrapper.debug()).toMatchSnapshot();
  });
});
