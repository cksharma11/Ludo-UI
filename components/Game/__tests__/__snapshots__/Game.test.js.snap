// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game should render properly 1`] = `
"<div className=\\"jsx-1772502628 body\\">
  <GameHeader title=\\"Ludo\\" playerName=\\"Chandan\\" />
  <section className=\\"jsx-1772502628 main_container\\">
    <section className=\\"jsx-1772502628 player_row\\">
      <Player name=\\"Arif\\" turn={true} coinData={{...}} />
      <CoinContainer playerId=\\"player_1\\" color=\\"red\\" deActiveCoins={{...}} />
      <CellRow className=\\"cell_row\\" cellId={1} color=\\"blue\\" containerClass=\\"\\" />
      <CoinContainer playerId=\\"player_2\\" color=\\"blue\\" deActiveCoins={{...}} />
      <Player name=\\"Chandan\\" turn={false} coinData={{...}} />
    </section>
    <div className=\\"jsx-1772502628 board_middle_row\\">
      <CellRow className=\\"cell_column\\" cellId={19} color=\\"red\\" containerClass=\\"cell_column_container\\" />
      <div className=\\"jsx-1772502628 clear_coins_container\\" />
      <CellRow className=\\"cell_column\\" cellId={37} color=\\"yellow\\" containerClass=\\"cell_column_container\\" />
    </div>
    <section className=\\"jsx-1772502628 player_row\\">
      <Player name=\\"Goarav\\" turn={false} coinData={{...}} />
      <CoinContainer playerId=\\"player_3\\" color=\\"green\\" deActiveCoins={{...}} />
      <CellRow className=\\"cell_row\\" cellId={55} color=\\"green\\" containerClass=\\"\\" />
      <CoinContainer playerId=\\"player_4\\" color=\\"yellow\\" deActiveCoins={{...}} />
      <Player name=\\"Pankaj\\" turn={false} coinData={{...}} />
    </section>
  </section>
  <JSXStyle id=\\"1772502628\\">
    .
    b
    o
    d
    y
    .
    j
    s
    x
    -
    1
    7
    7
    2
    5
    0
    2
    6
    2
    8
    {
    w
    i
    d
    t
    h
    :
    1
    0
    0
    %
    ;
    h
    e
    i
    g
    h
    t
    :
    1
    0
    0
    %
    ;
    p
    a
    d
    d
    i
    n
    g
    :
    0
    ;
    m
    a
    r
    g
    i
    n
    :
    0
    ;
    f
    o
    n
    t
    -
    f
    a
    m
    i
    l
    y
    :
    &#39;
    T
    r
    e
    b
    u
    c
    h
    e
    t
     
    M
    S
    &#39;
    ,
    &#39;
    L
    u
    c
    i
    d
    a
     
    S
    a
    n
    s
     
    U
    n
    i
    c
    o
    d
    e
    &#39;
    ,
    &#39;
    L
    u
    c
    i
    d
    a
     
    G
    r
    a
    n
    d
    e
    &#39;
    ,
     
    &#39;
    L
    u
    c
    i
    d
    a
     
    S
    a
    n
    s
    &#39;
    ,
    A
    r
    i
    a
    l
    ,
    s
    a
    n
    s
    -
    s
    e
    r
    i
    f
    ;
    }
    .
    m
    a
    i
    n
    _
    c
    o
    n
    t
    a
    i
    n
    e
    r
    .
    j
    s
    x
    -
    1
    7
    7
    2
    5
    0
    2
    6
    2
    8
    {
    d
    i
    s
    p
    l
    a
    y
    :
    -
    w
    e
    b
    k
    i
    t
    -
    b
    o
    x
    ;
    d
    i
    s
    p
    l
    a
    y
    :
    -
    w
    e
    b
    k
    i
    t
    -
    f
    l
    e
    x
    ;
    d
    i
    s
    p
    l
    a
    y
    :
    -
    m
    s
    -
    f
    l
    e
    x
    b
    o
    x
    ;
    d
    i
    s
    p
    l
    a
    y
    :
    f
    l
    e
    x
    ;
    p
    a
    d
    d
    i
    n
    g
    :
    1
    5
    p
    x
    ;
    -
    w
    e
    b
    k
    i
    t
    -
    a
    l
    i
    g
    n
    -
    i
    t
    e
    m
    s
    :
    c
    e
    n
    t
    e
    r
    ;
    -
    w
    e
    b
    k
    i
    t
    -
    b
    o
    x
    -
    a
    l
    i
    g
    n
    :
    c
    e
    n
    t
    e
    r
    ;
    -
    m
    s
    -
    f
    l
    e
    x
    -
    a
    l
    i
    g
    n
    :
    c
    e
    n
    t
    e
    r
    ;
    a
    l
    i
    g
    n
    -
    i
    t
    e
    m
    s
    :
    c
    e
    n
    t
    e
    r
    ;
    b
    a
    c
    k
    g
    r
    o
    u
    n
    d
    :
    #
    d
    f
    d
    f
    d
    f
    ;
    -
    w
    e
    b
    k
    i
    t
    -
    f
    l
    e
    x
    -
    d
    i
    r
    e
    c
    t
    i
    o
    n
    :
    c
    o
    l
    u
    m
    n
    ;
    -
    m
    s
    -
    f
    l
    e
    x
    -
    d
    i
    r
    e
    c
    t
    i
    o
    n
    :
    c
    o
    l
    u
    m
    n
    ;
    f
    l
    e
    x
    -
    d
    i
    r
    e
    c
    t
    i
    o
    n
    :
    c
    o
    l
    u
    m
    n
    ;
    }
    .
    p
    l
    a
    y
    e
    r
    _
    r
    o
    w
    .
    j
    s
    x
    -
    1
    7
    7
    2
    5
    0
    2
    6
    2
    8
    {
    d
    i
    s
    p
    l
    a
    y
    :
    -
    w
    e
    b
    k
    i
    t
    -
    b
    o
    x
    ;
    d
    i
    s
    p
    l
    a
    y
    :
    -
    w
    e
    b
    k
    i
    t
    -
    f
    l
    e
    x
    ;
    d
    i
    s
    p
    l
    a
    y
    :
    -
    m
    s
    -
    f
    l
    e
    x
    b
    o
    x
    ;
    d
    i
    s
    p
    l
    a
    y
    :
    f
    l
    e
    x
    ;
    }
    .
    b
    o
    a
    r
    d
    _
    m
    i
    d
    d
    l
    e
    _
    r
    o
    w
    .
    j
    s
    x
    -
    1
    7
    7
    2
    5
    0
    2
    6
    2
    8
    {
    d
    i
    s
    p
    l
    a
    y
    :
    -
    w
    e
    b
    k
    i
    t
    -
    b
    o
    x
    ;
    d
    i
    s
    p
    l
    a
    y
    :
    -
    w
    e
    b
    k
    i
    t
    -
    f
    l
    e
    x
    ;
    d
    i
    s
    p
    l
    a
    y
    :
    -
    m
    s
    -
    f
    l
    e
    x
    b
    o
    x
    ;
    d
    i
    s
    p
    l
    a
    y
    :
    f
    l
    e
    x
    ;
    w
    i
    d
    t
    h
    :
    6
    3
    0
    p
    x
    ;
    -
    w
    e
    b
    k
    i
    t
    -
    b
    o
    x
    -
    p
    a
    c
    k
    :
    j
    u
    s
    t
    i
    f
    y
    ;
    -
    w
    e
    b
    k
    i
    t
    -
    j
    u
    s
    t
    i
    f
    y
    -
    c
    o
    n
    t
    e
    n
    t
    :
    s
    p
    a
    c
    e
    -
    b
    e
    t
    w
    e
    e
    n
    ;
    -
    m
    s
    -
    f
    l
    e
    x
    -
    p
    a
    c
    k
    :
    j
    u
    s
    t
    i
    f
    y
    ;
    j
    u
    s
    t
    i
    f
    y
    -
    c
    o
    n
    t
    e
    n
    t
    :
    s
    p
    a
    c
    e
    -
    b
    e
    t
    w
    e
    e
    n
    ;
    }
    .
    c
    o
    i
    n
    _
    s
    y
    m
    b
    o
    l
    .
    j
    s
    x
    -
    1
    7
    7
    2
    5
    0
    2
    6
    2
    8
    {
    h
    e
    i
    g
    h
    t
    :
    5
    0
    %
    ;
    w
    i
    d
    t
    h
    :
    5
    0
    %
    ;
    -
    w
    e
    b
    k
    i
    t
    -
    a
    l
    i
    g
    n
    -
    s
    e
    l
    f
    :
    c
    e
    n
    t
    e
    r
    ;
    -
    m
    s
    -
    f
    l
    e
    x
    -
    i
    t
    e
    m
    -
    a
    l
    i
    g
    n
    :
    c
    e
    n
    t
    e
    r
    ;
    a
    l
    i
    g
    n
    -
    s
    e
    l
    f
    :
    c
    e
    n
    t
    e
    r
    ;
    b
    o
    r
    d
    e
    r
    :
    1
    p
    x
     
    s
    o
    l
    i
    d
     
    #
    f
    1
    b
    0
    9
    1
    e
    b
    ;
    b
    o
    r
    d
    e
    r
    -
    r
    a
    d
    i
    u
    s
    :
    5
    0
    %
    ;
    }
    .
    c
    l
    e
    a
    r
    _
    c
    o
    i
    n
    s
    _
    c
    o
    n
    t
    a
    i
    n
    e
    r
    .
    j
    s
    x
    -
    1
    7
    7
    2
    5
    0
    2
    6
    2
    8
    {
    w
    i
    d
    t
    h
    :
    0
    ;
    h
    e
    i
    g
    h
    t
    :
    0
    ;
    b
    o
    r
    d
    e
    r
    -
    t
    o
    p
    :
    6
    3
    p
    x
     
    s
    o
    l
    i
    d
     
    r
    g
    b
    (
    8
    4
    ,
    8
    4
    ,
    2
    3
    5
    )
    ;
    b
    o
    r
    d
    e
    r
    -
    b
    o
    t
    t
    o
    m
    :
    6
    3
    p
    x
     
    s
    o
    l
    i
    d
     
    r
    g
    b
    (
    1
    1
    2
    ,
    2
    3
    6
    ,
    1
    1
    2
    )
    ;
    b
    o
    r
    d
    e
    r
    -
    l
    e
    f
    t
    :
    6
    3
    p
    x
     
    s
    o
    l
    i
    d
     
    r
    g
    b
    (
    2
    3
    1
    ,
    9
    6
    ,
    9
    6
    )
    ;
    b
    o
    r
    d
    e
    r
    -
    r
    i
    g
    h
    t
    :
    6
    3
    p
    x
     
    s
    o
    l
    i
    d
     
    r
    g
    b
    (
    2
    3
    0
    ,
    2
    3
    0
    ,
    9
    1
    )
    ;
    }
  </JSXStyle>
</div>"
`;
